Operand::extract_field: only cast llval if it's a pointer and replace bitcast w/ pointercast.

luqmana opened the pull request:
Fixes #105439.

Also cc @erikdesjardins, looks like another place to cleanup as part of #105545

luqmana committed:
Operand::extract_field: only cast llval if it's a pointer and replace bitcast w/ pointercast.

luqmana committed:
Add test.

tmiasko commented on src/test/ui/simd/issue-105439.rs:
```diff
@@ -0,0 +1,14 @@
+// This is used to ICE with MIR inlining enabled due to an invalid bitcast.
+// run-pass
+// compile-flags: -O -Zmir-opt-level=3
```
Perhaps a smaller test would capture the essence of the issue without requiring MIR optimizations:

```rust
#![feature(repr_simd)]
#![feature(platform_intrinsics)]

#[allow(non_camel_case_types)]
#[derive(Clone, Copy)]
#[repr(simd)]
pub struct i32x4([i32; 4]);

pub fn f(a: i32x4) -> [i32; 4] {
    let b = a;
    b.0
}
```

```console
$ rustc --crate-type=lib a.rs -Zverify-llvm-ir
Invalid bitcast
  %4 = bitcast <4 x i32> %3 to [4 x i32]
LLVM ERROR: Broken module found, compilation aborted!
```

luqmana replied:
Sure, updated test.

luqmana committed:
Make test more targeted.

erikdesjardins commented on compiler/rustc_codegen_ssa/src/mir/operand.rs:
```diff
@@ -224,15 +225,24 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
                 *llval = bx.to_immediate(*llval, field);
                 // HACK(eddyb) have to bitcast pointers until LLVM removes pointee types.
-                *llval = bx.bitcast(*llval, bx.cx().immediate_backend_type(field));
+                let ty = bx.cx().immediate_backend_type(field);
+                if bx.type_kind(ty) == TypeKind::Pointer {
+                    *llval = bx.pointercast(*llval, ty);
+                }
```
This removes the illegal `bitcast <4 x i32> to [4 x i32]`, which is good--but it doesn't replace that illegal conversion with a legal one, so we end up returning a value of type `<4 x i32>` when extracting a field of type `[4 x i32]`.

The repro example happens to work anyways because the next instruction is a store, and both `store <4 x i32>` and `store [4 x i32]` have the same effect. But it won't work if we use that value by value. (This may never happen, because I don't think we ever pass LLVM array types by value, but it's technically incorrect.)

To handle this fully, I think you'd have to special-case this kind of newtype field extraction (self `Abi::Vector`, field `Abi::Aggregate`) and roundtrip the vector through memory (alloca, `store <4 x i32>`, `load [4 x i32]`).

(I thought there might be [endianness concerns](https://llvm.org/docs/LangRef.html#vector-type), but it seems like that's [only a problem](https://godbolt.org/z/18o6nEcv5) for not-a-multiple-of-i8 types like i4)

luqmana replied:
Good point! Split out the case for newtype of simd array

luqmana committed:
Don't bitcast aggregate field.

nagisa commented:
I had started reviewing it a while before, but clearly I’m not getting back to this anytime soon. I'll keep this tab open still, maybe I’ll find some time.

r? rust-lang/compiler

Nilstrieb commented on compiler/rustc_codegen_ssa/src/mir/operand.rs:
```diff
@@ -220,19 +221,53 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
                 // HACK(eddyb) have to bitcast pointers until LLVM removes pointee types.
-                *llval = bx.bitcast(*llval, bx.cx().immediate_backend_type(field));
+                let ty = bx.cx().immediate_backend_type(field);
+                if bx.type_kind(ty) == TypeKind::Pointer {
+                    *llval = bx.pointercast(*llval, ty);
+                } else {
+                    *llval = bx.bitcast(*llval, ty);
+                }
```
Maybe it makes sense to have a more generic cast method that does the check itself?

erikdesjardins replied:
We basically only generate pointer casts, so I don't think such an operation would be very widely used. (All existing usage of `bitcast` that I have seen was intended to convert between pointer types, and almost all can be removed after opaque pointers, e.g. https://github.com/rust-lang/rust/pull/105412/commits/68cabc0c87fe4f89ab3a6434b5adc7d5ddf77b87)

Excluding pointer types, the only other things that can be `bitcast`ed in LLVM are differently-typed vectors `<4 x i8> to <2 x i16>` and between vectors and integers `<4 x i8> to i32`. This would probably only be used in `transmute`. (We currently codegen such transmutes as roundtrips through memory: https://godbolt.org/z/M6fr1af8z.)

I don't think we support either of those operations as part of a field extraction, so I suspect the `else` here isn't necessary. @luqmana did you encounter code that still needs the `bitcast`?

luqmana replied:
I did not. Updated to just the pointercasts.

oli-obk commented:
considering how subtle this case is, what's your opinion on adding a codegen test?

erikdesjardins commented:
(I don't think you're asking me but) A codegen test could be useful, although the proper codegen is just a store/load which isn't too interesting.

Whether or not we add a codegen test, I'd suggest changing the existing test to `run-pass`, and add some additional code and assertions to make sure our codegen behaves properly at runtime. (e.g. generate an `i32x4` with some interesting values, call the function to convert it to `[i32; 4]`, and assert that the array elements are correct)

oli-obk commented:
@rustbot author

JohnCSimon commented:
@luqmana Ping from triage:  Can you post your status on this PR? This has sat idle for a few months.

luqmana committed:
Update test location.

luqmana committed:
Switch test back to run-pass.

luqmana committed:
Address review comments.

Remove bitcasts in OperandRef::extract_field; only pointercasts should
be needed.

luqmana commented:
Ok, rebased this and addressed review comments. Also switched the test back to run-pass:

```console
# Fails on latest nightly
$ rustc +nightly -vV
rustc 1.71.0-nightly (74c482104 2023-05-04)
binary: rustc
commit-hash: 74c4821045c68d42bb8b8a7c998bdb5c2a72bd0d
commit-date: 2023-05-04
host: aarch64-apple-darwin
release: 1.71.0-nightly
LLVM version: 16.0.2
$ rustc +nightly -O -Zverify-llvm-ir issue-105439.rs
Invalid bitcast
  %16 = bitcast <4 x i32> %15 to [4 x i32]
LLVM ERROR: Broken module found, compilation aborted!

# Passes with PR
$ rustc +stage1 -O -Zverify-llvm-ir issue-105439.rs
$ ./issue-105439 && echo $?
0
```



@rustbot ready

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit c7c042ad31766a6f70a2c850457f086cf6a03153 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk c7c042ad31766a6f70a2c850457f086cf6a03153 -->
<!-- homu: {"type":"Approved","sha":"c7c042ad31766a6f70a2c850457f086cf6a03153","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
